server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1654,10 +1655,12 @@ public void flush(boolean force, boolean waitIfOngoing) throws EngineException {
             }
             try {
                 // Only flush if (1) Lucene has uncommitted docs, or (2) forced by caller, or (3) the
-                // newly created commit points to a different translog generation (can free translog)
+                // newly created commit points to a different translog generation (can free translog),
+                // or (4) the processed local checkpoint is ahead of the local checkpoint tracked in the safe commit
                 boolean hasUncommittedChanges = indexWriter.hasUncommittedChanges();
                 boolean shouldPeriodicallyFlush = shouldPeriodicallyFlush();
-                if (hasUncommittedChanges || force || shouldPeriodicallyFlush) {
+                if (hasUncommittedChanges || force || shouldPeriodicallyFlush
+                        || getProcessedLocalCheckpoint() > getPersistedLocalCheckpoint()) {
```

Comment:
This is very subtle, as the persisted local checkpoint is driven mainly by Translog fsyncing (even though both local checkpoints are initialized from what's in the safe commit), i.e. getPersistedLocalCheckpoint() means the local checkpoint last persisted by the translog, not Lucene. I'm inclined to instead use `lastCommittedSegmentInfos.getUserData()...` here and look at the actually last Lucene-persisted local checkpoint.

Reply:
++ moved to reading it from the `lastCommittedSegmentInfos`